close-pull-requests/main.go

```diff
@@ -0,0 +1,289 @@
+		return nil, err
+	}
+
+	lastModif := lastHumanIssue
+	if lastInterestingEvent.After(*lastModif) {
+		lastModif = lastInterestingEvent
+	}
+
+	return lastModif, nil
+}
+
+func findLatestWarningComment(obj *github.MungeObject) *githubapi.IssueComment {
+	var lastFoundComment *githubapi.IssueComment
+
+	comments, err := obj.ListComments()
+	if err != nil {
+		return nil
+	}
+
+	for i := range comments {
+		comment := comments[i]
+		if !validIssueComment(comment) {
+			continue
+		}
+		if !isBotName(*comment.User.Login) {
+			continue
+		}
+
+		if !warningCommentRE.MatchString(*comment.Body) {
+			continue
+		}
+
+		if lastFoundComment == nil || lastFoundComment.CreatedAt.Before(*comment.UpdatedAt) {
+			if lastFoundComment != nil {
+				obj.DeleteComment(lastFoundComment)
+			}
+			lastFoundComment = &comment
+		}
+	}
+
+	return lastFoundComment
+}
+
+func durationToDays(duration time.Duration) string {
+	days := duration / day
+	dayString := "days"
+	if days == 1 || days == -1 {
+		dayString = "day"
+	}
+	return fmt.Sprintf("%d %s", days, dayString)
+}
+
+func closePullRequest(obj *github.MungeObject, inactiveFor time.Duration) {
+	comment := findLatestWarningComment(obj)
+	if comment != nil {
+		obj.DeleteComment(comment)
+	}
+
+	obj.WriteComment(fmt.Sprintf(closingComment, durationToDays(inactiveFor)))
+	obj.ClosePR()
+}
+
+func postWarningComment(obj *github.MungeObject, inactiveFor time.Duration, closeIn time.Duration) {
+	obj.WriteComment(fmt.Sprintf(
+		warningComment,
+		durationToDays(inactiveFor),
+		durationToDays(closeIn)))
+}
+
+func checkAndWarn(obj *github.MungeObject, reminder time.Duration, inactiveFor time.Duration, closeIn time.Duration) {
+	if closeIn < day {
+		// We are going to close the PR in less than a day. Too late to warn
+		return
+	}
+	comment := findLatestWarningComment(obj)
+	if comment == nil {
+		// We don't already have the comment. Post it
+		postWarningComment(obj, inactiveFor, closeIn)
+	} else if time.Since(*comment.UpdatedAt) > reminder {
+		// It's time to warn again
+		obj.DeleteComment(comment)
+		postWarningComment(obj, inactiveFor, closeIn)
+	} else {
+		// We already have a warning, and it's not expired. Do nothing
+	}
+}
+
+func processPullRequest(config *closerConfig, obj *github.MungeObject) error {
+	if !obj.IsPR() {
+		return nil
+	}
+
+	if obj.HasLabel(keepOpenLabel) {
+		return nil
+	}
+
+	lastModif, err := findLastModificationTime(obj)
+	if err != nil {
+		return err
+	}
+
+	closeIn := -time.Since(lastModif.Add(config.closeAfter))
+	inactiveFor := time.Since(*lastModif)
+	if closeIn <= 0 {
+		closePullRequest(obj, inactiveFor)
+	} else if closeIn <= config.warnBefore {
+		checkAndWarn(obj, config.warnReminder, inactiveFor, closeIn)
+	} else {
+		// Pull-request is active. Do nothing
+	}
+	return nil
+}
+
+func runProgram(config *closerConfig) error {
+	if err := config.PreExecute(); err != nil {
+		return err
+	}
+
+	return config.ForEachIssueDo(func(obj *github.MungeObject) error {
```

Comment:
See the design doc I sent which suggests a better way of doing this than iterating through every PR -- aka only search for PRs that have not been modified in a long time (to make stale/close) or issues with the stale label that have been modified recently (to remove the stale label)